### PR TITLE
8264768: JFR: Allow events to be printed to the log

### PR DESCRIPTION
**Pull request by egahlin**
Hi,

Could I have a review of change that allow events to be printed to the log. 

This feature is meant for debugging purposes and to simplify development of new events, both user-defined and JDK events. It's not suitable for production environments because 1) The amount of data that will printed in the default configuration 2) The overhead of formatting the output 3) Events may contain sensitive information that should not end up in a log file.

To print only user-defined events to standard out:
$  java -Xlog:jfr+event=debug -XX:StartFlightRecording ...

To print only JDK events to standard out:
$  java -Xlog:jfr+event+system=debug -XX:StartFlightRecording ...

To print both user-defined and JDK events to standard out:
$  java -Xlog:jfr+event*=debug -XX:StartFlightRecording ...

To print more than five frames in a stack trace:
$  java -Xlog:jfr+event*=trace -XX:StartFlightRecording ...

To print all events to standard out without decoration, similar to 'jfr print'
$  java -Xlog:jfr+event*=debug::none -XX:StartFlightRecording ...

Example outputs of -Xlog:jfr+event*=debug:

[2.444s][debug][jfr,system,event] jdk.CPULoad {
[2.444s][debug][jfr,system,event]   startTime = 15:32:09.605
[2.444s][debug][jfr,system,event]   jvmUser = 8,27%
[2.444s][debug][jfr,system,event]   jvmSystem = 0,68%
[2.444s][debug][jfr,system,event]   machineTotal = 12,78%
[2.444s][debug][jfr,system,event] }

[7.352s][debug][jfr,event       ] testing.FooBarBaz {
[7.352s][debug][jfr,event       ]   startTime = 15:33:11.526
[7.352s][debug][jfr,event       ]   duration = 1,00 s
[7.352s][debug][jfr,event       ]   eventThread = "main" (javaThreadId = 1)
[7.352s][debug][jfr,event       ]   stackTrace = [
[7.352s][debug][jfr,event       ]     Test.method3() line: 32
[7.352s][debug][jfr,event       ]     Test.method2() line: 24
[7.352s][debug][jfr,event       ]     Test.method1() line: 20
[7.352s][debug][jfr,event       ]     Test.main(String[]) line: 15
[7.352s][debug][jfr,event       ]     jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Method, Object, Object[])
[7.352s][debug][jfr,event       ]     ...
[7.352s][debug][jfr,event       ]   ]
[7.352s][debug][jfr,event       ] }

There is no good way to do filtering per event, since log tags needs to be known at compile time and JFR allows user-defined events to be registered at runtime. Maybe something can be invented that allows UL to accept some JFR specific syntax, i.e. -Xlog:jfr[jdk.ConcurrentModeFailure,jdk.AllocationRequiringGC], but it's outside the scope of this enhancement request. 

A custom .jfc file can always be created for the events you want to record, i.e. -Xlog:jfr+event*=debug -XX:StartFlightRecording:settings=custom.jfc

Testing: jdk/jdk/jfr

Thanks
Erik

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264768](https://bugs.openjdk.java.net/browse/JDK-8264768): JFR: Allow events to be printed to the log

### Reviewers
 * [Markus Grönlund](https://openjdk.java.net/census#mgronlun) (@mgronlun - **Reviewer**) ⚠️ Review applies to 7a2f5c38834fd51d72cebd421fb71b01006c002e

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3377/head:pull/3377` \
`$ git checkout pull/3377`

Update a local copy of the PR: \
`$ git checkout pull/3377` \
`$ git pull https://git.openjdk.java.net/jdk pull/3377/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3377`

View PR using the GUI difftool: \
`$ git pr show -t 3377`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3377.diff">https://git.openjdk.java.net/jdk/pull/3377.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/jfr/jni/jfrJavaSupport.cpp [hunk 1]
```diff
@@ -37,6 +37,7 @@
 #include "oops/klass.inline.hpp"
 #include "oops/oop.inline.hpp"
 #include "oops/objArrayKlass.hpp"
+#include "oops/objArrayOop.hpp"
 #include "oops/objArrayOop.inline.hpp"
```
Comment (stefank) [inline src/hotspot/share/jfr/jni/jfrJavaSupport.cpp:41]: We usually don't explicitly include the .hpp file when the .inline.hpp file is already included.

File: src/hotspot/share/jfr/utilities/jfrJavaLog.cpp [hunk 2]
```diff
@@ -28,7 +28,10 @@
 #include "jfr/utilities/jfrLogTagSets.hpp"
 #include "logging/log.hpp"
 #include "logging/logConfiguration.hpp"
+#include "logging/logMessage.hpp"
 #include "memory/resourceArea.hpp"
+#include "oops/objArrayOop.hpp"
+#include "oops/objArrayOop.inline.hpp"
```
Comment (stefank) [inline src/hotspot/share/jfr/utilities/jfrJavaLog.cpp:34]: Same here.

File: src/hotspot/share/jfr/jni/jfrJavaSupport.cpp (hunk 1 above)
Comment (stefank) [reply src/hotspot/share/jfr/jni/jfrJavaSupport.cpp:41]: (I also see that the lines many a few of the include  lines are not sorted)

**Review by stefank — COMMENTED** [review]
Nice feature!

Just peeked at the patch and saw some include nits.

File: src/hotspot/share/jfr/utilities/jfrJavaLog.cpp (hunk 2 above)
Comment (egahlin) [reply src/hotspot/share/jfr/utilities/jfrJavaLog.cpp:34]: I ran into issue with building (without precompiled headers on Linux), if I didn't include "#include "oops/objArrayOop.hpp" in jfrJavaLog.cpp. Maybe there is some better way around this.

Comment (stefank) [reply src/hotspot/share/jfr/utilities/jfrJavaLog.cpp:34]: I looked into this, and there is a circular dependency in objArrayOop.inline.hpp. I have created a PR to take care of that:
https://github.com/openjdk/jdk/pull/3431

Testing different build configurations with this now.

**Review by mgronlun — APPROVED** [review]
Looks good. Nice - convenient for development!

**Pull request closed without merging**